### ISSUE TITLE
CentOS 7.2 unshare(CLONE_NEWNS) fails

### ISSUE THREAD

**Issue by perj** [lxc/lxc#842]
I'm trying to create an unprivileged container as root on a CentOS 7.2 box, with user namespaces enabled in the kernel (`grubby --args=user_namespace.enable=1` + reboot).

Running `lxc-create -n test -t download` as root with these lines in lxc.conf:
```
lxc.id_map = u 0 9900000 100000
lxc.id_map = g 0 9900000 100000
```

The newuidmap and newgidmap binaries are not available on this machine (not distributed by CentOS).

There's some code I traced to lxc_usernsexec.c:
```
[pid 28970] execve("/bin/lxc-usernsexec", ["lxc-usernsexec", "-m", "u:0:9900000:100000", "-m", "g:0:9900000:100000", "-m", "u:100000:0:1", "-m", "g:100000:0:1", "--", "/usr/share/lxc/templates/lxc-dowload", "--path=/var/lib/lxc/test", "--name=test", "--rootfs=/var/lib/lxc/test/rootfs", "--mapped-uid", "100000", ...], [/* 23 vars */]) = 0
```

It first sets up the namespaces:
```
[pid 28973] unshare(CLONE_NEWNS|CLONE_NEWUSER) = 0 # child
...
[pid 28970] open("/proc/28973/uid_map", O_WRONLY|O_CREAT|O_TRUNC, 0666) = 5 # parent
[pid 28970] write(5, "0 9900000 100000\n100000 0 1\n", 28) = 28
```

Everything goes well here. But then there's a second call to unshare, and this one fails:

```
[pid 28973] unshare(CLONE_NEWNS)        = -1 EPERM (Operation not permitted) # child
```

lxc-create then aborts before I get to choose an image to download.

Perhaps this second call to CLONE_NEWNS isn't really needed? It'd be great if it could be avoided if so, since it seems everything works except for that.
Looking at the commit log this was the initial commit for lxc_usernsexec.c so that didn't tell me very much. I tried commenting out the second call to unshare, and as far as I can tell everything works as it should, running with uid 9900000 as root inside the container, but I don't really know this API well enough to make a definite call.

LXC version: 1.0.8 installed from EPEL. New lxc_usernsexec.c built from git master branch commit b450a79dd5cc2282c4d2ca56544378cc89ead44e

I know the kernel is probably acting in an unexpected way by denying the second call, it seems the RedHat kernel is a bit special in this regard.

**Comment by brauner** [comment]
Can you show the contents of the files `/etc/subuid` and `/etc/subgid`? They should contain the subuids and subgids you're trying to map your container to, i.e. an entry of the form: `USERNAME:9900000:100000`.

**Comment by perj** [comment]
I don't think those files are used running this as root? As far as I know they are for when I run lxc-create as a non-root user, which is not possible on a CentOS 7.2 kernel.
As mentioned I don't have the binaries `newuidmap` and `newgidmap` as of now, and I don't have lines defined for root in there. The writes to `/proc/<pid>/uid_map` and `/proc/<pid>/gid_map` are done by the `lxc_usernsexec` process.

Here's the current contents, regardless. Both files are identical:
```
pelle:100500000:100000
```

**Comment by brauner** [comment]
By specifying `lxc.id_map` you are establishing an id mapping. In short, this mapping takes care that the container runs unprivileged. This however, is orthogonal to whether you create it as root or as a normal user. Both, root and normal user can run unprivileged containers when a mapping is specified in the (default) configuration file. For this to work you will most likely need `newuidmap` and `newgidmap`.

If however, all you want is to create and run a privileged container then you should simply remove the two `lxc.id_map` lines from your (default) configuration file and try again.

**Comment by perj** [comment]
I want to create an unprivileged container. And I have, by commenting out the second call to `unshare` inside `lxc_usernsexec.c` I got it running. But I'm unsure if commenting out that call is safe and proper, and that's what this issue is about.

The `lxc_map_ids` function has two modes. One if `newuidmap` is installed (`use_shadow` is true), and one if it isn't. I'm using the second mode, where the function writes to the `uid_map` directly.

You are however correct that if it had been installed, then root would have needed lines in `/etc/subuid`. I've checked that.

**Comment by brauner** [comment]
Hm, I can't reproduce when installing a centos container, then installing lxc from epel and lxc_usernsexec.c from master and trying to create an unprivileged container. But this seems to be indeed a problem with red hat kernels: https://bugzilla.redhat.com/show_bug.cgi?id=1184294. I don't know if they backported the fix already.

**Comment by brauner** [comment]
Hm, wait that may not be immediately relevant to your problem.

**Comment by brauner** [comment]
Can you upload the strace as a .txt file. You should be able to just drag and drop it here.

**Comment by perj** [comment]
[lxc-create-strace.txt](https://github.com/lxc/lxc/files/142722/lxc-create-strace.txt)

The first unshare call is on line 600 and the second on 629. I commented out the second one that fails.

**Comment by brauner** [comment]
Does your kernel by any chance have a patch for turning of unprivileged `CLONE_NEWUSER`? Look at `/proc/sys/kernel/` for files named something with `userns`, `unprivileged`. If so, write `1` into that file and try again.

**Comment by perj** [comment]
Hmm, no, I can't find anything like that in there. It should be noted that this is a 3.10 kernel with some user namespace functionality backported, as is typical for how RedHat adds features in minor releases. `uname -r` is `3.10.0-327.10.1.el7.x86_64`.

It's `CLONE_NEWNS` that seems to require root (real root, not only user namespace root), `CLONE_NEWUSER`, and most of the rest `CLONE_NEW...` flags works as non-root.

**Comment by brauner** [comment]
Can you run:
```
unshare -fU -- echo bla
```
and
```
sudo unshare -fU -- echo bla
```

Do you receive an error and if so which?

**Comment by perj** [comment]
They both work without error:
```
$ unshare -fU -- echo bla
bla
$ sudo unshare -fU -- echo bla
bla
```

Adding the `m` flag only works for root though:
```
$ unshare -fUm -- echo bla
unshare: unshare failed: Operation not permitted
$ sudo unshare -fUm -- echo bla
bla
```

**Comment by brauner** [comment]
I setup a `KVM` for a centos 7.2 system to test. When I execute `lxc-create -n test -t download` I get the same unshare error about not being allowed to `CLONE_NEWNS`. This is a limitation of you kernel version and there is, to my knowledge, nothing we can do about it. Even when upgrading `shadow-utils` to version `2.2.x` you will face the same limitation.

**Comment by brauner** [comment]
And you can't remove the second call to `unshare()` in `lxc_usernsexec.c`. The uid/gid-mapping would not be successful in this case and you would end up with a privileged container contrary to your expectations. This is dangerous, don't do it.

**Comment by brauner** [comment]
If you don't have any follow-ups please don't forget to close the issue. :)

**Comment by perj** [comment]
Just remember that user namespaces are disabled by default, they have to be enabled by the kernel boot flag `user_namespace.enable=1`.

But ok, thanks for the definite answer, and for the thorough investigation. I have to admit I'm still fairly confused about why the uid/gid mapping would not work when it's the mount namespace call that I'm skipping, but I'll take your word for it.

**Comment by brauner** [comment]
Yes, I did exactly what you did, use `grubby` to append `user_namespace.enable=1` to grub; I can exactly reproduce your problem.

**Comment by brauner** [comment]
No, I was wrong about the `uid/gid` mapping not working I was thinking about another piece of code.

**Comment by brauner** [comment]
Maybe you can get your hands on a newer kernel and try again.

**Comment by popovich-sergei** [comment]
This issue is purely RHEL7/CentOS7 specific.

Currently user namespaces are in TECH PREVIEW phase in 7.2 branch and there is no guarantee that feature works correctly at all.

On the other hand, issue happened only when calling unshare(CLONE_NEWNS) from unprivileged process. In case of LXC this means calling unshare() after user uid/gid mapping setup and it does not relate to user namespaces at all.

Origin of the -EPERM is copy_mnt_ns() from fs/namespaces.c in Linux kernel tree. In upstream 3.10.x series there is no check for capability(CAP_SYS_ADMIN) (i.e. global privilege), where as in
RHEL/CentOS supplied kernels we have following:

```c
/* Unprivileged creation currently disabled in RHEL7 */
if (!capable(CAP_SYS_ADMIN))
    return ERR_PTR(-EPERM);
```

RH reasons on disabling unprivileged CLONE_NEWNS becoming much more cleaner from the following tread: [docker in OpenVZ](https://lists.openvz.org/pipermail/devel/2016-January/067586.html). By comparing Linux Kernel Upstream v3.10.y stable branch and RH kernel fs/namespaces.c we can easily find how much security related changes not in RH kernel. Thus keeping it disabled seems appropriate workaround for many potential security problems with unprivileged CLONE_NEWNS.

Also it seems that problem present only during LXC container creation: by creating unprivileged container from 4.4.x kernel and then starting it on 3.10.x from RH I'm able to run it successfuly (also unprivileged, as  seen from the initial user namespace).

The only way to fix LXC container creation from the stock RH kernels is to patch and rebuild kernel.

My version of [patch](https://github.com/lxc/lxc/files/314307/0001-mnt-Enable-unprivileged-namespace-creation-if-user-n.patch.gz) that works for me (tested on 3.10.0-327.18.2.el7.x86_64).

**Comment by hallyn** [comment]
Quoting Sergey Popovich (notifications@github.com):
> This issue is purely RHEL7/CentOS7 specific.
> 
> Currently user namespaces are in TECH PREVIEW phase in 7.2 branch and there is no guarantee that feature works correctly at all.
> 
> On the other hand, issue happened only when calling unshare(CLONE_NEWNS) from unprivileged process. In case of LXC this means calling unshare() after user uid/gid mapping setup and it does not relate to user namespaces at all.

Unprivileged users calling unshare very much relates to user
namespaces.  Without a private user namespace to limit the user's
ability to call setuid-root programs in the manipulated namespace,
unshare is not safe for unprivileged users.

> Origin of the -EPERM is copy_mnt_ns() from fs/namespaces.c in Linux kernel tree. In upstream 3.10.x series there is no check for capability(CAP_SYS_ADMIN) (i.e. global privilege), where as in
> RHEL/CentOS supplied kernels we have following:
> 
> ```c
> /* Unprivileged creation currently disabled in RHEL7 */
> if (!capable(CAP_SYS_ADMIN))
>     return ERR_PTR(-EPERM);
> ```
> 
> RH reasons on disabling unprivileged CLONE_NEWNS becoming much more cleaner from the following tread: [docker in OpenVZ](https://lists.openvz.org/pipermail/devel/2016-January/067586.html). By comparing Linux Kernel Upstream v3.10.y stable branch and RH kernel fs/namespaces.c we can easily find how much security related changes not in RH kernel. Thus keeping it disabled seems appropriate workaround for many potential security problems with unprivileged CLONE_NEWNS.
> 
> Also it seems that problem present only during LXC container creation: by creating unprivileged container from 4.4.x kernel and then starting it on 3.10.x from RH I'm able to run it successfuly (also unprivileged, as  seen from the initial user namespace).
> 
> The only way to fix LXC container creation from the stock RH kernels is to patch and rebuild kernel.
> 
> My version of [patch](https://github.com/lxc/lxc/files/314307/0001-mnt-Enable-unprivileged-namespace-creation-if-user-n.patch.gz) that works for me (tested on 3.10.0-327.18.2.el7.x86_64).

Without looking through all the rest of that kernel source to see what
checks are currently being made, i would recommend turning the capable
into a ns_capable.

**Comment by popovich-sergei** [comment]
> Unprivileged users calling unshare very much relates to user
namespaces.  Without a private user namespace to limit the user's
ability to call setuid-root programs in the manipulated namespace,
unshare is not safe for unprivileged users.

I did mean not related to the LXC unprivileged container creation issue in general, not relation between namespaces. But thanks you wery much for in-depth explanation. Thanks.

> Without looking through all the rest of that kernel source to see what checks are currently being made, i would recommend turning the capable into a ns_capable.

That was first thing I consider to do, but later I decided to try to keep RH change functionality as close as possible so make it in effect only when user_namespace.enable=1 (RH specific) is in effect.

Thanks.
